cmake/initialize-submodules.cmake

```diff
@@ -1,27 +1,89 @@
+      set(have_path FALSE)
+
       foreach(line ${file_lines})
-        if (${line} MATCHES "url =")
-          string(REGEX REPLACE "\\s*url = .*/(.*).git" "\\1" submodule "${line}")
-          string(STRIP "${submodule}" submodule)
-          if(NOT EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/${submodule}")
-            message(FATAL_ERROR "Submodule directory \"${CMAKE_CURRENT_SOURCE_DIR}/${submodule}\" does not exist")
+        if (NOT is_submodule)
+          string(COMPARE EQUAL ${line} "[submodule \"${module_ref_name}\"]" is_submodule)
+          if (is_submodule)
+            message(STATUS "\"${module_name}\" is a submodule of this project.")
+            continue()
           endif()
-          # Initialize submodule if it hasn't already been cloned
-          if(NOT EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/${submodule}/.git")
-            message(STATUS "Initialize ${submodule} submodule")
-            execute_process(COMMAND ${GIT_EXECUTABLE} submodule update --init --recursive "${CMAKE_CURRENT_SOURCE_DIR}/${submodule}"
-              WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
-              RESULT_VARIABLE GIT_SUBMOD_RESULT)
-            if(NOT GIT_SUBMOD_RESULT EQUAL "0")
-              message(FATAL_ERROR "git submodule update --init --recursive ${CMAKE_CURRENT_SOURCE_DIR}/${submodule} failed with ${GIT_SUBMOD_RESULT}, please checkout submodules")
-            endif()
+        endif()
+
+        if (is_submodule AND (NOT have_path))
+          string(FIND ${line} "path =" pos)
+          set(have_path NOT(pos EQUALS -1))
+          if (have_path)
+            string(REPLACE "path =" "" submodule_path ${line})
+            string(STRIP "${submodule_path}" submodule_path)
+            set(submodule_path "${PROJECT_SOURCE_DIR}/${submodule_path}")
+            continue()
           endif()
         endif()
       endforeach()
+
+      if(is_submodule)
+        if(have_path)
+          message(STATUS "Updating submodule \"${module_name}\" at ${submodule_path}")
+          execute_process(COMMAND ${GIT_EXECUTABLE} submodule update --init --recursive "${submodule_path}"
+                  WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+                  RESULT_VARIABLE GIT_SUBMOD_RESULT)
+          if(GIT_SUBMOD_RESULT EQUAL "0")
+            message(STATUS "Successfully updated submodule \"${module_name}\"")
+          else()
+            message(FATAL_ERROR "Unable to update submodule \"${module_name}\"")
+          endif()
+        endif()
+      else()
+        message(STATUS "\"${module_name}\" is not a submodule of this project")
+      endif()
+
+    endif()
+  endif()
+
+  if (TARGET ${module_name})
+    message(STATUS "Submodule \"${module_name}\" is a target.")
+  else()
+    if(EXISTS "${module_path}")
+      message(STATUS "Adding subdirectory ${module_path} to this project")
+      add_subdirectory(${module_path})
     endif()
   endif()
-endmacro()
\ No newline at end of file
+
+  message("")
```

Comment:
???

Reply:
Give this version a try in a few situations and feel free to change the messaging however you prefer.